### PR TITLE
Content-defined-chunking and progress bar

### PR DESCRIPTION
**Pull request by borismattijssen**
This PR adds content-defined-chunking to uploading and downloading data.
It also includes a progress bar that shows the up/download progress.
This progress bar can be disabled if the `--json-format` flag is used.

### CODE REVIEW

File: command/upload.go [hunk 1]
```diff
@@ -99,12 +115,172 @@ func (cmd *Upload) DoRun(args []string) (err error) {
...
+		rel, err := filepath.Rel(dir, path)
+		if err != nil {
+			return errors.Wrapf(err, "failed to determine path '%s' relative to '%s'", path, dir)
+		}
+
+		f, err := os.Open(path)
+		if err != nil {
+			return errors.Wrapf(err, "failed to open file '%s'", rel)
+		}
+
```
Comment (advanderveer) [inline command/upload.go]: (Minor) f.Close() should be placed here

File: command/upload.go [hunk 2]
```diff
@@ -99,12 +115,172 @@ func (cmd *Upload) DoRun(args []string) (err error) {
...
+		err = tw.WriteHeader(&tar.Header{
+			Name:    rel,
+			Mode:    int64(fi.Mode()),
+			ModTime: fi.ModTime(),
```
Comment (advanderveer) [inline command/upload.go]: Do we want ModTimes() to invalidate the chunks?

File: command/upload.go [hunk 3]
```diff
@@ -99,12 +115,172 @@ func (cmd *Upload) DoRun(args []string) (err error) {
...
+		HandleError(errors.Wrapf(err, "failed to tar '%s'", dataPath), cmd.opts.VerboseOutput)
+	}
+
+	pw.Close()
+	err = <-doneCh
+	<-progressBarDoneCh
+	if err != nil {
+		HandleError(errors.Wrapf(err, "failed to upload '%s'", dataPath), cmd.opts.VerboseOutput)
+	}
+
```
Comment (advanderveer) [inline command/upload.go]: Would it be possible to report the number of deduplicated bytes? the user might be confused on why it was so quick?

File: nerd/data/dataset.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+//Metadata describes a dataset. It contains a header with different properties of the dataset and a
+//KeyReadWriter which is used to keep track of the list of Keys (chunks) of the dataset.
+type Metadata struct {
+	Header *MetadataHeader
+	KeyReadWriter
+}
+
+//MetadataHeader is the header of the dataset metadata.
+type MetadataHeader struct {
+	Created time.Time `json:"created_at"`
```
Comment (advanderveer) [inline nerd/data/dataset.go]: these times are probably set by local machine's clock? UTC? its probably ok, we just need to be careful that they are not used for any logic (clock skews and what not)

Comment (borismattijssen) [inline nerd/data/dataset.go]: True thing! Maybe I should just leave it out for now? It's not needed for any functionality at the moment.

File: command/upload.go (hunk 2 above)
Comment (borismattijssen) [inline command/upload.go]: I think we do. The header will only be a few chunks. Invalidating these chunks is worth the up-to-date information in the archive header.

File: command/upload.go (hunk 3 above)
Comment (borismattijssen) [inline command/upload.go]: I like the idea. Let's add it later as a feature. I created #59 for this.

**Comment by borismattijssen** [comment]
Addressed comments

File: command/upload.go (hunk 2 above)
Comment (advanderveer) [inline command/upload.go]: 👍

File: nerd/data/dataset.go (hunk 4 above)
Comment (advanderveer) [inline nerd/data/dataset.go]: Ok, think is is part of a larger topic/brainstorm. Especially around backwards (in)compatibility. I created: https://github.com/nerdalize/nerd/issues/60

**Comment by advanderveer** [comment]
I'm just gonna merge things, awesome work!

**Pull request merged**